app/src/main/java/org/cgutman/usbip/usb/XferUtils.java

```diff
@@ -52,9 +52,10 @@ public static int doBulkTransfer(UsbDeviceConnection devConn, UsbEndpoint endpoi
 			
 			bytesTransferred += res;
 			
-			if (res < endpoint.getMaxPacketSize()) {
+			if (res != endpoint.getMaxPacketSize()) {
```

Comment:
Hmm, I wonder if this is needed or if the `while` condition is sufficient. Does it work properly if you remove this check entirely?

Reply:
First of all the request comes with buff.length (expected) size of 65536. The only reasonable interpretation is maximum size. So we can't check buff.length vs. transferred size. Yubikey is sending back a certificate which is by a variable size.
Second, I know this code is by definition wrong, because I can manipulate certificate to be exactly max packet size.

Now back to your question.
If I take the condition out it will terminate with timeout and return error.
Perhaps if there is nothing more, it would return right away. Nop, it is a real timeout with 1000ms.
I am reading the server part of the linux driver to figure out how it is done. First impression, things depend on transfer flags.